lib/validations.js

```diff
@@ -340,7 +340,20 @@ function validateCustom(attr, conf, err, options, done) {
     done = options;
     options = {};
   }
-  conf.customValidator.call(this, err, done);
+
+  if (conf.customValidator.length <= 1) {
+    conf.customValidator.call(this, options)
+      .then((isValid) => {
+        if (!isValid) throw new Error();
```

Comment:
AFAICT, there are several outcomes of a custom validator:

 1. `err()` is called - an error message is added by the validation framework
 2. `err(false)` is called - the validated object is marked as invalid, but no error message is added by the validation framework. See https://github.com/strongloop/loopback-datasource-juggler/pull/1229/files#diff-a1a277b8e694ce811867b2ac3e064ff8L638
 3. `done()` is called - the property value is valid
 4. `done(truthy)` is called - same as calling `err()`

I would like to preserve these options in the Promise mode too, i.e.
 1. Indicate that the validation failed - items 1 and 4 above
 2. Indicate that the validation failed and the error message was already inserted - item 2 above
 3. Indicate that the validation passed

I think the cleanest solution is to expect the promise to resolve with a structured object instead of a single true/false value, although we can support boolean resolution to make simple cases simple to implement.

```js
// the value is valid
return Promise.resolve(); 

// the value is invalid, a message should be added by the framework
return Promise.reject();
return Promise.reject(err);
throw unhandledError;

// the value is invalid, a message was already added by the validator
return Promise.reject({ addMessage: false });
``` 

Here is a code snippet illustrating a possible implementation of my proposal:


```js
var result = conf.customValidator(...);
if (result.then && typeof result.then === 'function') {
  result.then(
    function onSuccess() { done(); },
    function onFailure(reason) {
      if (reason instanceof Error) {
        done(err);
      } else if (reason && reason.addMessage !== false) {
        err(false);
      } else {
        err();
      }
    })
}
```

Thoughts?